differ/differ.go

```diff
@@ -155,21 +167,122 @@ func (entry oneEntry) Diff(other oneEntry) (int, bool) {
 	} else if entry.OpCode != other.OpCode {
 		return 0, false
 	} else if entry.OpCode == gomemcached.UPR_MUTATION {
-		if entry.RevId != other.RevId {
+		entryRevId := entry.GetRevId()
+		otherRevId := other.GetRevId()
+		if entryRevId != otherRevId && entryRevId != 0 && otherRevId != 0 { //Darshan : RevId comparison should be revisited
 			return 0, false
-		} else if entry.Cas != other.Cas {
+		} else if entry.GetVersion() != other.GetVersion() {
 			return 0, false
 		} else if entry.Flags != other.Flags {
 			return 0, false
 		} else if !shaCompare(entry.BodyHash, other.BodyHash) {
 			return 0, false
 		} else if entry.Datatype != other.Datatype {
 			return 0, false
+		} else if !compareHlv(entry.Hlv, other.Hlv, entry.Cas, other.Cas, entry.ImportCas, other.ImportCas, entry.bucketUUID, other.bucketUUID) {
+			return 0, false
 		}
 	}
 	return 0, true
 }
 
+func compareHlv(hlv1, hlv2 *hlv.HLV, cas1, cas2, importCas1, importCas2 uint64, bucketUUID1, bucketUUID2 hlv.DocumentSourceId) bool {
+	//if both(source and target) have no HLV we return true
+	if hlv1 == nil && hlv2 == nil {
+		return true
+	} else if hlv1 != nil && hlv2 == nil { //this is the case where the document is replicated from target to source
+		if hlv1.GetCvSrc() != bucketUUID2 || hlv1.GetCvVer() != cas2 {
+			return false
+		} else {
+			hlv2, _ = hlv.NewHLV(bucketUUID2, cas2, 0, "", 0, nil, nil)
+		}
+	} else if hlv1 == nil && hlv2 != nil { //this is the case where the document is replicated from source to target
+		if hlv2.GetCvSrc() != bucketUUID1 || hlv2.GetCvVer() != cas1 {
+			return false
+		} else {
+			hlv1, _ = hlv.NewHLV(bucketUUID1, cas1, 0, "", 0, nil, nil)
+		}
+	}
+	return actualCompare(hlv1, hlv2, cas1, cas2)
+}
+
+func actualCompare(item1 *hlv.HLV, item2 *hlv.HLV, item1Cas uint64, item2Cas uint64) bool {
+	// the HLVs that this function recieves is up to date
+	if item1.GetCvCas() != item2.GetCvCas() {
+		return false
+	} else if item1.GetCvSrc() != item2.GetCvSrc() || item1.GetCvVer() != item2.GetCvVer() {
+		return false
+	} else if !comparePv(item1.GetPV(), item2.GetPV(), item1Cas, item2Cas) {
+		return false
+	}
+	return true
+}
+
+func comparePv(Pv1 hlv.VersionsMap, Pv2 hlv.VersionsMap, cas1 uint64, cas2 uint64) bool {
```

Comment:
This function seems to be a prime candidate to write unit test cases for.
Try writing some unit tests and test boundary cases for this in `differ_test.go` as part of this commit

Reply:
Sure, I'll write some unit tests